bus-mapping/src/operation.rs

```diff
@@ -0,0 +1,369 @@
+            _ => Err(Error::InvalidOpConversion),
+        }
+    }
+}
+
+/// Doc
+#[derive(Debug, Clone)]
+pub(crate) enum Operation {
+    /// Doc
+    Stack(StackOp),
+    /// Doc
+    Memory(MemoryOp),
+    /// Doc
+    Storage(StorageOp),
+}
+
+impl From<&StackOp> for Operation {
+    fn from(op: &StackOp) -> Self {
+        Operation::Stack(op.clone())
+    }
+}
+
+impl From<StackOp> for Operation {
+    fn from(op: StackOp) -> Self {
+        Operation::Stack(op)
+    }
+}
+
+impl From<&MemoryOp> for Operation {
+    fn from(op: &MemoryOp) -> Self {
+        Operation::Memory(op.clone())
+    }
+}
+
+impl From<MemoryOp> for Operation {
+    fn from(op: MemoryOp) -> Self {
+        Operation::Memory(op)
+    }
+}
+
+impl From<&StorageOp> for Operation {
+    fn from(op: &StorageOp) -> Self {
+        Operation::Storage(op.clone())
+    }
+}
+
+impl From<StorageOp> for Operation {
+    fn from(op: StorageOp) -> Self {
+        Operation::Storage(op)
+    }
+}
+
+impl PartialEq for Operation {
+    fn eq(&self, other: &Operation) -> bool {
+        match (self, other) {
+            (Operation::Stack(stack_op_1), Operation::Stack(stack_op_2)) => {
+                stack_op_1.eq(stack_op_2)
+            }
+            (Operation::Memory(memory_op_1), Operation::Memory(memory_op_2)) => {
+                memory_op_1.eq(memory_op_2)
+            }
+            (Operation::Storage(storage_op_1), Operation::Storage(storage_op_2)) => {
+                storage_op_1.eq(storage_op_2)
+            }
+            _ => false,
+        }
+    }
+}
+
+impl Eq for Operation {}
+
+impl PartialOrd for Operation {
+    fn partial_cmp(&self, other: &Operation) -> Option<Ordering> {
+        match (self, other) {
+            (Operation::Stack(stack_op_1), Operation::Stack(stack_op_2)) => {
+                stack_op_1.partial_cmp(stack_op_2)
+            }
+            (Operation::Memory(memory_op_1), Operation::Memory(memory_op_2)) => {
+                memory_op_1.partial_cmp(memory_op_2)
+            }
+            (Operation::Storage(storage_op_1), Operation::Storage(storage_op_2)) => {
+                storage_op_1.partial_cmp(storage_op_2)
+            }
+            _ => None,
+        }
+    }
+}
+
+impl Operation {
+    pub const fn target(&self) -> Target {
+        match self {
+            Operation::Memory(_) => Target::Memory,
+            Operation::Stack(_) => Target::Stack,
+            Operation::Storage(_) => Target::Storage,
+        }
+    }
+
+    pub const fn is_stack(&self) -> bool {
+        matches!(*self, Operation::Stack(_))
+    }
+
+    pub const fn is_memory(&self) -> bool {
+        matches!(*self, Operation::Memory(_))
+    }
+
+    pub const fn is_storage(&self) -> bool {
+        matches!(*self, Operation::Storage(_))
+    }
+
+    pub fn into_stack_unchecked(self) -> StackOp {
+        match self {
+            Operation::Stack(stack_op) => unsafe { std::mem::transmute(stack_op) },
```

Comment:
Do we really need `unsafe transmute` here? I just test without this and still pass test

Reply:
As you can see in https://doc.rust-lang.org/std/mem/fn.transmute.html. `transmute` is an unsafe operation.

I want to bench at some point the difference between the `transmute` impl and a `TryFrom`. If the difference is not significant, and the resulting code does not look really bad, it woubl make sense to make the change.

But this is the only way I saw to downcast an enum into it's variant when the variants have different types without introducing `Error`,

Also note that I have commented out the `#![deny(unsafe_code)]` in `lib.rs`

Reply:
side question: why do we need `into_*_unchecked()` methods when we already have `TryFrom<Operation>` implemented for each of the variants?